Dependencies/xr/Source/ARKit/XR.mm

```diff
@@ -0,0 +1,838 @@
+            MetalDevice = id<MTLDevice>(graphicsContext);
+            sessionDelegate = [[SessionDelegate new]init:&ActiveFrameViews metalContext:MetalDevice];
+            session.delegate = sessionDelegate;
+            [session runWithConfiguration:configuration];
+
+            // build pipeline
+            NSError* error;
+            id<MTLLibrary> lib = CompileShader(MetalDevice, shaderSource);
+            id<MTLFunction> vertexFunction = [lib newFunctionWithName:@"vertexShader"];
+            id<MTLFunction> fragmentFunction = [lib newFunctionWithName:@"fragmentShader"];
+
+            // Configure a pipeline descriptor that is used to create a pipeline state.
+            MTLRenderPipelineDescriptor *pipelineStateDescriptor = [[MTLRenderPipelineDescriptor alloc] init];
+            pipelineStateDescriptor.label = @"XR Pipeline";
+            pipelineStateDescriptor.vertexFunction = vertexFunction;
+            pipelineStateDescriptor.fragmentFunction = fragmentFunction;
+            pipelineStateDescriptor.colorAttachments[0].pixelFormat = MetalLayer.pixelFormat;
+
+            pipelineState = [MetalDevice newRenderPipelineStateWithDescriptor:pipelineStateDescriptor error:&error];
+            if (!pipelineState)
+            {
+                NSLog(@"Failed to create pipeline state: %@", error);
+            }
+            
+            [pipelineStateDescriptor release];
+            commandQueue = [MetalDevice newCommandQueue];
+            
+            // default fov values to avoid NaN at startup
+            auto& frameView = ActiveFrameViews[0];
+            frameView.FieldOfView.AngleDown = -(frameView.FieldOfView.AngleUp = 0.5);
+            frameView.FieldOfView.AngleLeft = -(frameView.FieldOfView.AngleRight = 0.5);
+        }
+
+        ~Impl()
+        {
+            if (ActiveFrameViews[0].ColorTexturePointer != nil)
+            {
+                id<MTLTexture> oldColorTexture = reinterpret_cast<id<MTLTexture>>(ActiveFrameViews[0].ColorTexturePointer);
+                [oldColorTexture setPurgeableState:MTLPurgeableStateEmpty];
+                [oldColorTexture release];
+                ActiveFrameViews[0].ColorTexturePointer = nil;
+            }
+            
+            if (ActiveFrameViews[0].DepthTexturePointer != nil)
+            {
+                id<MTLTexture> oldDepthTexture = reinterpret_cast<id<MTLTexture>>(ActiveFrameViews[0].DepthTexturePointer);
+                [oldDepthTexture setPurgeableState:MTLPurgeableStateEmpty];
+                [oldDepthTexture release];
+                ActiveFrameViews[0].DepthTexturePointer = nil;
+            }
+
+            [session pause];
+            session.delegate = nil;
+            [sessionDelegate release];
+            [pipelineState release];
+            dispatch_sync(dispatch_get_main_queue(), ^{
+                [xrView removeFromSuperview]; });
+            [xrView releaseDrawables];
+            [xrView dealloc];
+            xrView = nil;
```

Comment:
Do you not have to release the `session`?

Reply:
I am using a static session & config here that is persistent across sessions.  I'm looking into seeing if that was not actually necessary, and if I can just go back to having it owned by the instance itself and then I would need to call release here.

Reply:
Confirmed going back to releasing the session after every call would leak memory across sessions.  I think there are some resources that are not tied to the lifetime of the session that might need to be cleaned up separately.  Going to leave it as is for now.